sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala

```diff
@@ -95,12 +95,32 @@ private[sql] case class AggregateExpression2(
   override def toString: String = s"(${aggregateFunction},mode=$mode,isDistinct=$isDistinct)"
 }
 
-abstract class AggregateFunction2
-  extends Expression with ImplicitCastInputTypes {
+sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes {
```

Comment:
Logically, AggregateFunction2 is a union-type, the least-upper-bound of InterpretedAggregateFunction and ExpressionAggregateFunction (the temporary name which I'm using for expression-based aggregates).